### PR TITLE
mac-capture: show actual windows in Window Capture sources

### LINKED ISSUE #3146: Icons/system windows show as capture targets for Window Capture on macOS
When using Window Capture on macOS, several junk items appear, clogging the window capture list. 
These appear to be simply icons when captured, and should not be shown in the window list to avoid
clutter and confusion with end users. Examples:

- [Google Chrome] Item-0
- [OBS] Item-0
- [OBS] (all main OBS windows/settings windows)
- [Window Server] Backstop Menubar
- [Window Server] Desktop
- [Window Server] Menubar
- [TextInputMenuAgent] Item-0
- [Spotlight] Item-0

## Platform

Operating system and version: macOS 10.15.5 Catalina
OBS Studio version: 25.0.8

## Expected Behavior
System windows should not be displayed in window capture.
We currently have a blacklist for these items on Windows OS,
and this should be expanded to include items on macOS.

## Current Behavior
Window capture lists system windows not designed for capture.

## Steps to Reproduce
1. Open OBS
2. Add Window Capture
3. Check the list and observe several nonsense and Item-0 options in the list that are not actual windows.

## Additional information
N/A

### PR DESCRIPTION
**Pull request by timmiehaha**
### Description
This pull request addresses the issue found [here](https://github.com/obsproject/obs-studio/issues/3146) where sources inside Windows Capture show system/icon windows as well. Upon investigating the issue, it appears that the problem is inside `plugins/mac-capture/window-utils.m`:
```
NSArray *enumerate_windows(void)
{
	NSArray *arr = (NSArray *)CGWindowListCopyWindowInfo(
		kCGWindowListOptionOnScreenOnly, kCGNullWindowID);
	[arr autorelease];

	return [arr sortedArrayUsingComparator:win_info_cmp];
}
```

The `CGWindowListCopyWindowInfo(option, relativeToWindow)` returns every single window that is onscreen. It just so happens that the icons such as those in the menu bar

image

are also considered windows, which have different `kCGWindowLayer` levels. The ones we're concerned with are found at the standard window layer level, which is zero. In other words, we want to filter out the windows returned from `CGWindowListCopyWindowInfo` to only give us ones where their `kCGWindowLayer == 0`. The below snippet of code has been added to serve as the extra filtering step:
```
NSArray *filter_nonzero_kcgwindowlayer_sources(NSArray *windows_arr)
{
	NSPredicate *pred =
		[NSPredicate predicateWithFormat:@"(kCGWindowLayer == 0)"];
	NSArray *new_windows_arr =
		[windows_arr filteredArrayUsingPredicate:pred];

	return new_windows_arr;
}
```

### Motivation and Context
This pull request addresses the issue found [here](https://github.com/obsproject/obs-studio/issues/3146) where sources inside Windows Capture show system/icon windows as well. These only cause clutter with their large numbers and we presumably only want actual windows to show, so there is extra logic added to the code to clean up the windows.

### How Has This Been Tested?

I tested this on macOS Catalina 10.15.4. More importantly, I tested what would happen to existing "garbage" windows if this change is made. For example, I added `SystemUIServer - BatteryExtra` as a source and applied the code changes.
image
Upon restarting OBS, the window no longer showed up as a valid window. This would certainly affect those who have added sources like `SystemUIServer - BatteryExtra`, but it's nothing that would cause the program to crash. For reference, this is the new list of windows that appears when selecting a source:

image

As for how this could potentially affect other areas of code, it's pretty self-contained inside of `window-utils.m`. The only code that calls `enumerate_windows()` is code that needs to list out window candidates for a Window Capture source.

### Types of changes

- Bug fix (non-breaking change which fixes an issue)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

Fixes #3146

### CODE REVIEW

File: plugins/mac-capture/window-utils.m [hunk 1]
```diff
@@ -48,7 +55,6 @@ bool find_window(cocoa_window_t cw, obs_data_t *settings, bool force)
 
 	if (!cw->window_name.length && !cw->owner_name.length)
 		goto invalid_name;
-
```
Comment (WizardCM) [inline plugins/mac-capture/window-utils.m:51]: Please remove this line change from the diff.

**Review by WizardCM — COMMENTED** [review]
Thank you for looking into this!

File: plugins/mac-capture/window-utils.m (hunk 1 above)
Comment (timmiehaha) [reply plugins/mac-capture/window-utils.m]: Whoops, don't know how that got there. Removed it and re-ran clang after not realizing I was forgetting the `-i` flag... everything should be good now

**Comment by timmiehaha** [comment]
@RytoEX Thank you for the pointers. I've modified the commit message to now say `mac-capture: filter non-trivial windows` and added `Fixes #3146` to the end of the PR description.

File: plugins/mac-capture/window-utils.m [hunk 2]
```diff
@@ -23,14 +23,25 @@
...
+		[windows_arr filteredArrayUsingPredicate:pred];
+
+	return new_windows_arr;
+}
+
 NSArray *enumerate_windows(void)
 {
 	NSArray *arr = (NSArray *)CGWindowListCopyWindowInfo(
 		kCGWindowListOptionOnScreenOnly, kCGNullWindowID);
+	NSArray *filtered_arr = filter_nonzero_kcgwindowlayer_sources(arr);
```
Comment (RytoEX) [inline plugins/mac-capture/window-utils.m:40]: `arr` is autoreleased below.  Should `filtered_arr` also be autoreleased?

Comment (timmiehaha) [reply plugins/mac-capture/window-utils.m:40]: @RytoEX In Objective-C only objects that we own need to be released. According to [official documentation](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/MemoryMgmt/Articles/mmRules.html), this consists of objects we create using `new`, `init`, `copy`, or `mutableCopy`. We do not perform any of these actions to create `filtered_arr`, so we do not need to release it. In fact, calling either `[filtered_arr release]` or `[filtered_arr autorelease]` will attempt to free memory that's not there, resulting in OBS crashing because of a double free scenario.

I'm not too familiar with the internals of `CGWindowListCopyWindowInfo` but I'm guessing there's a `new` or `init` involved in its returned `NSArray`, which may be why it needs to be marked with `autorelease`.

Comment (RytoEX) [reply plugins/mac-capture/window-utils.m:40]: Perfectly valid to explain that much to someone who doesn't develop on macOS.  I understand the situation now.

**Review by RytoEX — APPROVED** [review]
From my perspective (code style and commit guidelines), this seems fine to me.  The code itself seems fine, but I'd like to see someone on our end with macOS build and test.

**Comment by WizardCM** [comment]
Heads up: due to the issue mentioned in #3528 this PR has been reverted for now. If possible, please submit a fresh PR with edge cases covered.
